releng_sop/koji_sign.py

```diff
@@ -0,0 +1,715 @@
+
+        :param tag_name: A koji tag name
+        :type  tag_name: str
+        :param inherit: Follow tag inheritance
+        :type  inherit: bool=False
+        :param packages: Filter RPMs by provided packages
+        :type  packages: list=None
+        :return: List of koji rpm_info dictionaries
+        :rtype:  list
+        """
+        result = []
+
+        if not packages:
+            packages = [None]
+
+        for package in set(packages):
+            rpm_info_list, build_info_list = self.koji_session.listTaggedRPMS(tag_name, latest=True, inherit=inherit, package=package, rpmsigs=False)
+            builds_by_id = {}
+            for build_info in build_info_list:
+                build_info["name"] = build_info["package_name"]
+                builds_by_id[build_info["id"]] = build_info
+            for rpm_info in rpm_info_list:
+                rpm_info["build"] = builds_by_id[rpm_info["build_id"]]
+            result.extend(rpm_info_list)
+        return result
+
+    def get_build_rpms(self, build_list):
+        """
+        Return rpm_info list for specified builds.
+
+        :param build_list: List of koji builds (NVRs)
+        :type  build_list: list
+        :return: List of koji rpm_info dictionaries
+        :rtype:  list
+        """
+        result = []
+        for build in build_list:
+            build_info = self.koji_session.getBuild(build, strict=True)
+            rpm_info_list = self.koji_session.listRPMs(buildID=build_info["id"])
+            for rpm_info in rpm_info_list:
+                rpm_info["build"] = build_info
+            result.extend(rpm_info_list)
+        return result
+
+    def filter_rpm_info_list_by_packages(self, rpm_info_list, package_list):
+        """
+        Return rpm_info list filtered by specified packages.
+
+        :param rpm_info_list: List of koji rpm_info dictionaries
+        :type  rpm_info_list: list
+        :param package_list: List of koji packages (SRPM names)
+        :type  package_list: list
+        :return: List of koji rpm_info dictionaries
+        :rtype:  list
+        """
+        # convert to set, list lookups are slow
+        package_list = set(package_list)
+
+        result = []
+        for rpm_info in rpm_info_list:
+            if rpm_info["build"]["name"] not in package_list:
+                continue
+            result.append(rpm_info)
+        return result
+
+    def get_rpm_sig_dict(self, rpm_info_list):
+        """
+        Read information about cached signatures from koji.
+
+        :param rpm_info_list: List of koji rpm_info dictionaries
+        :type  rpm_info_list: list
+        :return: {rpm_id: {sigkey: sighash}}
+        :rtype:  dict
+        """
+        result = {}
+
+        self.koji_session.multicall = True
+        for rpm_info in rpm_info_list:
+            self.koji_session.queryRPMSigs(rpm_info["id"])
+        data = self.koji_session.multiCall()
+
+        for [sig_list] in data:
```

Comment:
I added strict=True argument to raise exceptions on xmlrpc errors.
Doesn't fix the problem entirely, but it's good enough.